### PR TITLE
make some typedlist C-APIs public

### PR DESCRIPTION
**Pull request by dlee992**
Hi. I make some typedlist C-APIs public, since I want to use typedlist C-APIs in my C++ code, which will be called from jitted numba code, via `intrinsic`, `lower_builtin`, and etc.

At least, (I feel) I need `new`, `free`, `get_length`, `setitem`, and `getitem` APIs. I hope this PR won't bother other parts in numba or the design principle of Numba. If not, please shout.

I mark this PR as a draft, when I finish my extension and test it, then I will mark this as an official PR.

### CODE REVIEW

**Comment by gmarkall** [comment]
The consensus from the dev meeting on 2022-10-18 was that it's OK to make these APIs public, as long as it's documented that they're not supported / stable, and the interfaces or implementation could be changed in future without notice. So with some documentation of them not being supported / stable, this PR should be OK to merge.

**Comment by gmarkall** [comment]
Notes: https://hackmd.io/SXbcwPbCSwKgo4i52XH5zw

**Comment by gmarkall** [comment]
Sorry, I hit the wrong button and closed the PR.

**Comment by stuartarchibald** [comment]
> The consensus from the dev meeting on 2022-10-18 was that it's OK to make these APIs public, as long as it's documented that they're not supported / stable, and the interfaces or implementation could be changed in future without notice. So with some documentation of them not being supported / stable, this PR should be OK to merge.

I think "_changed_ **OR** _removed_ in future without notice" may better cover the potential liability associated with their use?

**Comment by dlee992** [comment]
Thanks, I've already made some changes in docs/header comments/codes, following your suggestions. Can begin to review.

File: docs/source/reference/pysupported.rst [hunk 1]
```diff
@@ -552,6 +552,11 @@ Typed List
   functionality or suffer from unexpectedly bad performance, please report
   this, ideally by opening an issue on the Numba issue tracker.
 
+.. warning::
+  Most interfaces listed in ``numba/cext/listobject.h`` are exported,
+  but they're not supported/stable, and these interfaces or underlying
+  implementations could be changed or removed in future without notice.
+
```
Comment (gmarkall) [inline docs/source/reference/pysupported.rst:559]: I think this note is a little out-of-place in the section of the manual that discusses support for Python features. Given that the same warning / disclaimer is present in `listobject.h`, I'd suggest just removing this change.

```suggestion
```

**Review by gmarkall — CHANGES_REQUESTED** [review]
This looks good to me (once the modification to pysupported.rst is removed), but given the nature of the changes and prior discussion, someone else should take a look too.

File: docs/source/reference/pysupported.rst (hunk 1 above)
Comment (dlee992) [reply docs/source/reference/pysupported.rst]: Okay, solved as you requested.

**Review by gmarkall — APPROVED** [review]
Thanks for the update - this looks good to me (though still should have a second look from someone else).

File: numba/cext/listobject.h [hunk 2]
```diff
@@ -4,7 +4,10 @@
  *
  * https://github.com/python/cpython/blob/51ddab8dae056867f3595ab3400bffc93f67c8d4/Include/listobject.h
  *
- *
+ * WARNING:
+ * Most interfaces listed here are exported, but they're not supported/stable,
+ * and these interfaces or underlying implementations could be changed or
+ * removed in future without notice.
```
Comment (stuartarchibald) [inline numba/cext/listobject.h:10]:
```suggestion
 * Most interfaces listed here are exported (global), but they are not
 * supported, stable, or part of Numba's public API. These interfaces and their
 * underlying implementations may be changed or removed in future without
 * notice.
```
suggest this to note that the symbols are global, and to make it explicit that they are not part of Numba's public API.

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch @dlee992. I've suggested a minor change to the warning text in an attempt to be more explicit with respect to a couple of points.

The only other thought I have is whether the exported symbols should have `protected` visibility just for additional safety (though this may be excessive). Do you have a view on this @gmarkall ?

**Comment by gmarkall** [comment]
> The only other thought I have is whether the exported symbols should have `protected` visibility just for additional safety (though this may be excessive). Do you have a view on this @gmarkall ?

It's not clear to me how that would make anything safer - can you elaborate?

**Comment by stuartarchibald** [comment]
> > The only other thought I have is whether the exported symbols should have `protected` visibility just for additional safety (though this may be excessive). Do you have a view on this @gmarkall ?
> 
> It's not clear to me how that would make anything safer - can you elaborate?

From OOB discussion, conclusion is that it probably makes insufficient difference. 

However, as part of this discussion, it was noted that the "default" behaviour in relation to symbol export varies across platforms and that it would be a good idea to ensure that symbols are exported consistently. For example, I think this patch as of ac31d64 will result in some typed list symbols being `global` on unix systems, but on windows I think they will need explicitly exporting to achieve the same. Thoughts?

**Comment by gmarkall** [comment]
I think the default visibility on OS X / Linux was already fine - on Windows I think `__declspec(dllexport)` is needed: https://learn.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-declspec-dllexport?view=msvc-170

**Comment by dlee992** [comment]
> I think the default visibility on OS X / Linux was already fine - on Windows I think `__declspec(dllexport)` is needed: https://learn.microsoft.com/en-us/cpp/build/exporting-from-a-dll-using-declspec-dllexport?view=msvc-170

Cool, I should revise it again based on this info.

**Review by gmarkall — DISMISSED** [review]
Thanks for the revisions - this looks good to me.

**Comment by gmarkall** [comment]
@stuartarchibald Are you happy with this PR at this point?

**Comment by stuartarchibald** [comment]
> @stuartarchibald Are you happy with this PR at this point?

I think what is now present makes the exporting of declared symbols as globally visible consistent across platforms, thanks for addressing this @dlee992.

File: numba/cext/cext.h [hunk 3]
```diff
@@ -9,6 +9,10 @@
 #define NUMBA_EXPORT_FUNC(_rettype) VISIBILITY_HIDDEN _rettype
 #define NUMBA_EXPORT_DATA(_vartype) VISIBILITY_HIDDEN _vartype
 
+/* Define symbols in this C module, and export them outside the
+   shared library in various platforms, including Windows. */
```
Comment (stuartarchibald) [inline numba/cext/cext.h:13]:
```suggestion
/* Use to declare a symbol as exported (global). */
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the updates @dlee992, one minor suggestion RE commentary for the C macro else looks good.

File: numba/cext/cext.h (hunk 3 above)
Comment (dlee992) [reply numba/cext/cext.h]: Thanks for the review!

**Review by gmarkall — APPROVED** [review]
Following the small comment update, I still approve.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes. I manually checked on a linux system that the supposedly exported symbols are indeed global in the `_helperlib.<version>.so`.

**Pull request merged**